pkg/ingester/wal.go

```diff
@@ -384,35 +383,89 @@ func recoverFromWAL(ingester *Ingester) (err error) {
-		start := time.Now()
-		if err := processCheckpoint(lastCheckpointDir, userStates, nWorkers, stateCache, seriesCache); err != nil {
-			return err
-		}
-		elapsed := time.Since(start)
-		level.Info(util.Logger).Log("msg", "recovered from checkpoint", "time", elapsed.String())
-	} else {
-		level.Info(util.Logger).Log("msg", "no checkpoint found")
+	level.Info(util.Logger).Log("msg", "recovering from checkpoint")
+	start := time.Now()
+	userStates, idx, err := processCheckpointWithRepair(walDir, ingester, nWorkers, stateCache, seriesCache)
+	if err != nil {
+		return err
 	}
+	elapsed := time.Since(start)
+	level.Info(util.Logger).Log("msg", "recovered from checkpoint", "time", elapsed.String())
+	defer func() {
+		if returnErr == nil {
+			ingester.userStatesMtx.Lock()
+			ingester.userStates = userStates
+			ingester.userStatesMtx.Unlock()
+		}
+	}()
 
 	if segExists, err := segmentsExist(walDir); err == nil && !segExists {
 		level.Info(util.Logger).Log("msg", "no segments found, skipping recover from segments")
 		return nil
 	}
 
 	level.Info(util.Logger).Log("msg", "recovering from WAL", "dir", walDir, "start_segment", idx)
-	start := time.Now()
-	if err := processWAL(walDir, idx, userStates, nWorkers, stateCache, seriesCache); err != nil {
+	start = time.Now()
+	if err := processWALWithRepair(walDir, idx, userStates, nWorkers, stateCache, seriesCache, ingester.metrics); err != nil {
 		return err
 	}
-	elapsed := time.Since(start)
+	elapsed = time.Since(start)
 	level.Info(util.Logger).Log("msg", "recovered from WAL", "time", elapsed.String())
 
 	return nil
 }
 
+func processCheckpointWithRepair(walDir string, ingester *Ingester, nWorkers int,
+	stateCache []map[string]*userState, seriesCache []map[string]map[uint64]*memorySeries) (*userStates, int, error) {
+
+	// Use a local userStates, so we don't need to worry about locking.
+	userStates := newUserStates(ingester.limiter, ingester.cfg, ingester.metrics)
+
+	lastCheckpointDir, idx, err := lastCheckpoint(walDir)
+	if err != nil {
+		return nil, -1, err
+	}
+	if idx < 0 {
+		level.Info(util.Logger).Log("msg", "no checkpoint found")
+		return userStates, -1, nil
+	}
+
+	level.Info(util.Logger).Log("msg", fmt.Sprintf("recovering from %s", lastCheckpointDir))
+
+	err = processCheckpoint(lastCheckpointDir, userStates, nWorkers, stateCache, seriesCache)
+	if err == nil {
+		return userStates, idx, nil
+	}
+
+	// We don't call repair on checkpoint as losing even a single record is like losing the entire data of a series.
+	// We try recovering from the older checkpoint instead.
+	ingester.metrics.walCorruptionsTotal.Inc()
+	level.Error(util.Logger).Log("msg", "checkpoint recovery failed, deleting this checkpoint and trying to recover from old checkpoint", "err", err)
+
+	// Deleting this checkpoint to try the previous checkpoint.
+	if err := os.RemoveAll(lastCheckpointDir); err != nil {
+		return nil, -1, err
```

Comment:
We may wrap this `err` to add a message about what operation we were doing, something like:
`errors.Wrapf(err, "unable to delete checkpoint directory %s", lastCheckpointDir`